src/utils.ts

```diff
@@ -655,3 +660,27 @@ export function fillErrorDetails(properties: IExecutionResultTelemetryProperties
         properties.exceptionId = e.id.toString();
     }
 }
+
+export function makeUnique<T>(elements: T[]): T[] {
+    return Array.from(new Set(elements));
+}
+
+/**
+ * adaptToSingleIntoToMulti transforms a function that takes a single parameter, to a version that takes an array and applies the function to each parameter.
+ * Some usage examples:
+ *   public readonly getBPsFromIDs = adaptToSingleIntoToMulti(this, this.getBPFromID);
+ *   public readonly toStackFrames = asyncAdaptToSingleIntoToMulti(this, this.toStackFrame);
+ *   public readonly toSourceTrees = asyncAdaptToSingleIntoToMulti(this, this.toSourceTree);
+ *   public readonly toBPRecipiesStatus = asyncAdaptToSingleIntoToMulti(this, this.toBPRecipieStatus);
+ */
+export function adaptToSingleIntoToMulti<T, R>(thisObject: object, toSingle: (single: T) => R): (multi: T[]) => R[] {
```

Comment:
This is confusing, I don't understand why a caller wouldn't just call map themselves. I don't think this makes code simpler or easier to follow.

Reply:
I agree it's simpler and "easier to follow" to write the full method code each time. The trade-off I'm doing here is to avoid writing the same code over and over again, which I think is more error-prone than just writing the code once, and reusing it in this way.

Reply:
What does it eliminate though, one call to 'map'? To me it doesn't seem worth the mental load of introducing a new concept just to save a few characters.

Reply:
Currently the code is:
```
    public readonly toStackFrames = asyncAdaptToSingleIntoToMulti(this, this.toStackFrame);
    public readonly toSourceTrees = asyncAdaptToSingleIntoToMulti(this, this.toSourceTree);
    public readonly toBPRecipiesStatus = asyncAdaptToSingleIntoToMulti(this, this.toBPRecipieStatus);

```
Without that the code would be:
```
    public toStackFrames(rows: IStackTracePresentationRow[]): Promise<DebugProtocol.StackFrame[]> {
        return asyncMap(rows, row => this.toStackFrame(row));
    }

```
and the same code for the other 2 methods...

Reply:
Honestly that looks perfectly fine to me. It's one less concept for someone to understand. I would probably skip the helper and just write 

`const frames = rows.map(r => this.toStackFrame(r))` whenever I need to do this conversion, and that doesn't seem like a big deal to me, it's totally transparent.

But it's not a big deal, I won't hold up the PR on it.